pyscf/pbc/pwscf/krks.py

```diff
@@ -0,0 +1,496 @@
+    if spin != 0:
+        nspin = len(C_ks)
+        assert nspin > 0
+    else:
+        nspin = 1
+        C_ks = [C_ks]
+        mocc_ks = [mocc_ks]
+    outshape = (nspin, nrho, np.prod(mesh))
+    rhovec_R = np.ndarray(outshape, buffer=out)
+    if nrho > 0:
+        for s in range(nspin):
+            rhovec_R[s, 0] = mf.with_jk.get_rho_R(
+                C_ks[s], mocc_ks[s], mesh=mesh, Gv=Gv
+            )
+    if nrho > 1:
+        for s in range(nspin):
+            rho_G = tools.fft(rhovec_R[s, 0], mesh)
+            for v in range(3):
+                drho_G = 1j * Gv[:, v] * rho_G
+                rhovec_R[s, v + 1] = tools.ifft(drho_G, mesh).real
+    if nrho > 4:
+        for s in range(nspin):
+            dC_ks = [np.empty_like(C_k) for C_k in C_ks[s]]
+            rhovec_R[s, 4] = 0
+            const = 1j * np.sqrt(0.5)
+            for v in range(3):
+                for k, C_k in enumerate(C_ks[s]):
+                    if mf.with_jk.basis_ks is None:
+                        ikgv = const * (mf.kpts[k][v] + Gv[:, v])
+                    else:
+                        ikgv = const * mf.with_jk.basis_ks[k].Gk[:, v]
+                    dC_ks[k][:] = ikgv * C_k
+                rhovec_R[s, 4] += mf.with_jk.get_rho_R(
+                    dC_ks, mocc_ks[s], mesh=mesh, Gv=Gv
+                )
+    if spin == 0:
+        rhovec_R = rhovec_R[0]
+    return rhovec_R
+
+
+def apply_vxc_kpt(mf, C_k, kpt, vxc_R, vtau_R=None, mesh=None, Gv=None,
+                  C_k_R=None, comp=None, basis=None):
+    """
+    Apply the XC potential to the bands C_k at a given kpt.
+    """
+    cell = mf.cell
+    if mesh is None: mesh = mf.wf_mesh
+    if Gv is None: Gv = cell.get_Gv(mesh)
+    if comp is not None:
+        vxc_R = vxc_R[comp]
+        if vtau_R is not None:
+            vtau_R = vtau_R[comp]
+    apply_j_kpt = mf.with_jk.apply_j_kpt
+    Cbar_k = apply_j_kpt(C_k, mesh, vxc_R, C_k_R=C_k_R, basis=basis)
+    if vtau_R is not None:
+        const = 1j * np.sqrt(0.5)
+        dC_k = np.empty_like(C_k)
+        for v in range(3):
+            if mf.with_jk.basis_ks is None:
+                ikgv = const * (kpt[v] + Gv[:, v])
+            else:
+                ikgv = const * basis.Gk[:, v]
+            dC_k[:] = ikgv * C_k
+            dC_k[:] = apply_j_kpt(dC_k, mesh, vtau_R, basis=basis)
+            Cbar_k[:] += ikgv.conj() * dC_k
+    return Cbar_k
+
+
+def eval_xc(mf, xc_code, rhovec_R, xctype):
+    if rhovec_R.ndim == 2:
+        spin = 0
+    else:
+        assert rhovec_R.ndim == 3
+        spin = 1
+    exc_R, vxcvec_R = mf._numint.eval_xc_eff(xc_code, rhovec_R, deriv=1,
+                                             xctype=xctype)[:2]
+    dv = mf.cell.vol / exc_R.size
+    if spin == 0:
+        vxcvec_R = vxcvec_R[None, ...]
+        rho_R = rhovec_R[0]
+        rhovec_R = rhovec_R.view()[None, ...]
+    else:
+        rho_R = rhovec_R[:, 0].sum(0)
+    exc = dv * exc_R.dot(rho_R)
+    return exc, vxcvec_R
```

Comment:
Is there a benefit to scale the exc value by the integral weight, here? exc and vxc have different "unit", seems misleading.